File: src/slave.js
#!/usr/bin/env node
const WebSocket = require('ws');
let robot;
try {
  robot = require('robotjs');
} catch (e) {
  console.error('\n[slave] Error: `robotjs` is not installed or failed to load.');
  console.error('[slave] On macOS you may need Xcode Command Line Tools and to build native modules.');
  console.error('[slave] Suggested steps:');
  console.error('  1) Install Xcode CLI: `xcode-select --install`');
  console.error('  2) Install node-gyp prerequisites (Python, etc.) and ensure `node` and `npm` are up-to-date.');
  console.error('  3) From the project directory run: `npm install --build-from-source robotjs`');
  console.error('If you prefer avoiding native modules, I can adapt this project to use `cliclick` (brew install cliclick) as a fallback.');
  process.exit(1);
}

const args = process.argv.slice(2);
const port = parseInt(args[0], 10) || 8080;

const wss = new WebSocket.Server({ port });
let masterSocket = null;
let controlling = false;

console.log(`[slave] starting WebSocket server on port ${port}`);

wss.on('connection', socket => {
  console.log('[slave] master connected');
  masterSocket = socket;
  socket.on('message', onMessage);
  socket.on('close', () => {
    console.log('[slave] master disconnected');
    masterSocket = null;
    controlling = false;
  });
});

function onMessage(raw) {
  let msg;
  try {
    msg = JSON.parse(raw);
  } catch (e) {
    return;
  }

  switch (msg.type) {
    case 'grab':
      controlling = true;
      console.log('[slave] master grabbed control');
      break;
    case 'mousemove':
      if (!controlling) return;
      const pos = robot.getMousePos();
      robot.moveMouse(pos.x + (msg.dx || 0), pos.y + (msg.dy || 0));
      break;
    case 'mouseclick':
      if (!controlling) return;
      const btn = mapButton(msg.button);
      if (msg.down) robot.mouseToggle('down', btn); else robot.mouseToggle('up', btn);
      break;
    case 'keydown':
      if (!controlling) return;
      // raw forwarding: best-effort; mapping may be required per-OS
      // If you need reliable keyboard forwarding, we can send characters instead of raw codes.
      if (typeof msg.keycode !== 'undefined') {
        try { robot.keyToggle(String(msg.keycode), 'down'); } catch (e) { }
      }
      break;
    case 'keyup':
      if (!controlling) return;
      if (typeof msg.keycode !== 'undefined') {
        try { robot.keyToggle(String(msg.keycode), 'up'); } catch (e) { }
      }
      break;
    default:
      break;
  }
}

function mapButton(b) {
  // iohook uses 1=left, 2=right, 3=middle sometimes; robotjs expects 'left'/'right'/'middle'
  if (b === 2) return 'right';
  if (b === 3) return 'middle';
  return 'left';
}

// Poll to detect left-screen edge on slave to return control
setInterval(() => {
  if (!controlling || !masterSocket || masterSocket.readyState !== WebSocket.OPEN) return;
  const pos = robot.getMousePos();
  if (pos.x <= 0) {
    // return control
    try {
      masterSocket.send(JSON.stringify({ type: 'release' }));
      controlling = false;
      console.log('[slave] released control back to master');
    } catch (e) {
      // ignore
    }
  }
}, 50);

File: python/slave.py
#!/usr/bin/env python3
"""Slave WebSocket server: applies received mouse/keyboard events locally.

Usage: python3 python/slave.py [port]
"""
import asyncio
import json
import argparse
import sys
import traceback

try:
    from pynput import mouse, keyboard
except Exception:
    print("[slave] Error: `pynput` is required. Install with: pip install pynput pyautogui websockets")
    raise

try:
    import pyautogui
except Exception:
    pyautogui = None

import websockets


class SlaveServer:
    def __init__(self, port):
        self.port = port
        self.mouse_ctrl = mouse.Controller()
        self.keyboard_ctrl = keyboard.Controller()
        self.controlling = False
        self.ws = None

    async def handler(self, websocket, path):
        print(f"[slave] master connected")
        self.ws = websocket
        try:
            async for raw in websocket:
                try:
                    msg = json.loads(raw)
                except Exception:
                    continue

                try:
                    t = msg.get('type')
                    if t == 'grab':
                        self.controlling = True
                        print('[slave] control acquired from master')
                    elif t == 'mousemove' and self.controlling:
                        dx = msg.get('dx', 0) or 0
                        dy = msg.get('dy', 0) or 0
                        x, y = self.mouse_ctrl.position
                        try:
                            self.mouse_ctrl.position = (int(x + dx), int(y + dy))
                        except Exception:
                            pass
                    elif t == 'mouseclick' and self.controlling:
                        btn = msg.get('button', 'left')
                        down = msg.get('down', True)
                        # map string to Button
                        btn_map = {
                            'Button.left': mouse.Button.left,
                            'Button.right': mouse.Button.right,
                            'Button.middle': mouse.Button.middle,
                            'left': mouse.Button.left,
                            'right': mouse.Button.right,
                            'middle': mouse.Button.middle
                        }
                        b = btn_map.get(btn, mouse.Button.left)
                        if down:
                            self.mouse_ctrl.press(b)
                        else:
                            self.mouse_ctrl.release(b)
                    elif t == 'keydown' and self.controlling:
                        k = msg.get('key')
                        self._press_key(k, down=True)
                    elif t == 'keyup' and self.controlling:
                        k = msg.get('key')
                        self._press_key(k, down=False)
                except Exception:
                    print('[slave] Exception while handling message:')
                    traceback.print_exc()
                    # don't re-raise; send error info back to client optionally
                    try:
                        await websocket.send(json.dumps({"type": "error", "message": "internal server error"}))
                    except Exception:
                        pass

        except websockets.ConnectionClosed:
            print('[slave] master disconnected')
        finally:
            self.ws = None
            self.controlling = False

    def _press_key(self, keystr, down=True):
        if keystr is None:
            return
        try:
            if keystr.startswith('Key.'):
                name = keystr.split('.', 1)[1]
                k = getattr(keyboard.Key, name)
                if down:
                    self.keyboard_ctrl.press(k)
                else:
                    self.keyboard_ctrl.release(k)
            else:
                # single character
                if down:
                    self.keyboard_ctrl.press(keystr)
                else:
                    self.keyboard_ctrl.release(keystr)
        except Exception:
            pass

    async def edge_watcher(self):
        # watch local mouse; if at left edge while controlling, release back to master
        while True:
            await asyncio.sleep(0.05)
            if not self.controlling or self.ws is None or self.ws.closed:
                continue
            x, y = self.mouse_ctrl.position
            if x <= 0:
                try:
                    await self.ws.send(json.dumps({"type": "release"}))
                    self.controlling = False
                    print('[slave] released control back to master')
                except Exception:
                    pass

    async def run(self):
        server = await websockets.serve(self.handler, '0.0.0.0', self.port)
        print(f"[slave] WebSocket server listening on :{self.port}")
        await self.edge_watcher()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('port', nargs='?', default=8080, type=int)
    args = parser.parse_args()
    srv = SlaveServer(args.port)
    try:
        asyncio.run(srv.run())
    except KeyboardInterrupt:
        print('[slave] exiting')


if __name__ == '__main__':
    main()
